tests/Test.ADAL.NET.Unit/InstanceDiscoveryTests.cs

```diff
@@ -157,5 +157,59 @@ public void TestInitialize()
             Assert.AreEqual("login.microsoftonline.com", entry2.PreferredNetwork);
             Assert.AreEqual(1, HttpMessageHandlerFactory.MockHandlersCount()); // Still 1 mock response remaining, so no new request was attempted
         }
+
+        [TestMethod]
+        [TestCategory("InstanceDiscoveryTests")]
+        public async Task TestInstanceDiscovery_WhenMetadataIsReturned_ShouldUsePreferredNetwork()
+        {
+            string host = "login.windows.net";
+            string preferredNetwork = "login.microsoftonline.com";
+
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
+            {
+                Method = HttpMethod.Get,
+                Url = $"https://{host}/common/discovery/instance",
+                ResponseMessage = new HttpResponseMessage(HttpStatusCode.OK)
+                {
+                    Content = new StringContent(
+                        @"{
+                        ""tenant_discovery_endpoint"" : ""https://login.microsoftonline.com/v1/.well-known/openid-configuration"",
+                        ""metadata"": [
+                            {
+                            ""preferred_network"": ""login.microsoftonline.com"",
+                            ""preferred_cache"": ""login.windows.net"",
+                            ""aliases"": [""login.microsoftonline.com"", ""login.windows.net"", ""sts.microsoft.com""]
+                            }
+                        ]
+                        }"
+                    )
+                }
+            });
+
+            var authenticator = new Authenticator($"https://{host}/contoso.com/", false);
+            await authenticator.UpdateFromTemplateAsync(new CallState(Guid.NewGuid()));
+
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler()
+            {
+                Method = HttpMethod.Post,
+                Url = $"https://{preferredNetwork}/contoso.com/oauth2/token", // This validates the token request is sending to expected host
+                ResponseMessage = new HttpResponseMessage(HttpStatusCode.OK)
+                {
+                    Content = new StringContent("{\"token_type\":\"Bearer\",\"expires_in\":\"3599\",\"access_token\":\"some-token\"}")
+                }
+            });
+
+            var privateObject = new PrivateObject(new AcquireTokenForClientHandler(new RequestData
+            {
+                Authenticator = authenticator,
+                Resource = "resource1",
+                ClientKey = new ClientKey(new ClientCredential("client1", "something")),
+                SubjectType = TokenSubjectType.Client,
+                ExtendedLifeTimeEnabled = false
+            }));
+            await (Task) privateObject.Invoke("SendTokenRequestAsync");
```

Comment:
why is this method being invoke via reflection? you could run runHandlerAsync(); instead

Reply:
Assuming you mean the `RunAsync()` method [defined here](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/9073fe37b6cb9f3d02436d0ce8103345c9e8a29e/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs#L117), yes I actually considered it too, and that would also work.

I prefer to not use it, because `RunAsync()`, which is basically the only available public methods in the `AcquireTokenHandlerBase` family, indirectly wraps the entire 400+ lines AcquireTokenHandlerBase behaviors: PreRunAsync(), PostRunAsync(), AcquireTokenUsingBroker(), CheckAndAcquireTokenUsingBroker(), SendTokenRequestAsync(), SendHttpMessageAsync(), to name a few. That means any changes in those helpers in the future could potentially break this test case. Now, given the fact that this PR actually only implements the new behavior inside SendHttpMessageAsync(), it seems more desirable to minimize blast radius to the scope of adjustments, which in this case happens to be a private method.

Whether or how we should unit test private methods is an inspiring topic in itself. Feel free to read [this stackoverflow question, its answers AND its comments](https://stackoverflow.com/questions/9122708/unit-testing-private-methods-in-c-sharp).

Reply:
The method is never called outside the pipeline that you described. this should be tested as a integration test from top to bottom. your test approach would be more effective if this functionality was in the AdalHttpClient as suggested below. then you could write tests hitting the adalhttpclient in various degrees

Reply:
Based on the recently discovered issues caused by test cases unnecessarily coupling each other, it seems like the **UNIT** tests should better stay away from "this should be tested as a integration test from top to bottom" approach. If every unit test case would indirectly cover (and therefore be potentially affected by) more than necessary scope, it won't be scalable in the long term. (We can talk in person.)

Reply:
Like I said before, refactoring the change into AdalHttpClient would warrant for unit tests (still mocked for network calls) as that class gets hit from multiple points in the code. Do you agree?